### PR TITLE
Fix warnings with default configure

### LINKED ISSUE #1252: [next] build very noisy with gcc 8.1.1
Most are from `-Wnull-dereference`

See https://omadesala.alfa1024.info/~bout-test/2018-09-05_17:18/build.log

There are some others as well.

Good that `-Werror` was not included.

### PR DESCRIPTION
**Pull request by ZedThree**
Fixes #1252 

- Adds `getBase` and `isEnabled` as protected methods to `Output`, and makes `ConditionalOutput` a friend class. This simplifies ConditionalOutput a bit by removing the need for `base_is_cond` and `dynamic_cast`. Also adds explicit nullptr checks at at levels. Should be ok, as output is not a big performance hit
- Pulls the loop over the `region` in `checkData` that appears at `CHECK > 2` into a separate (detail) function. This avoids some really ugly workarounds to put `UNUSED` on `region` only at certain `CHECK` levels
- Uses a temporary reference in `BoutMesh::pack/unpack_data` to avoid a null dereference warning. I'm not really sure it "fixes" it so much as hides it from the compiler. It does have the advantage that it makes the code a bit more readable though.

The last remaining warning is due to this: https://github.com/boutproject/BOUT-dev/blob/master/src/mesh/boundary_standard.cxx#L88

Should we always do this check? It shouldn't cost anything (we're already doing a jump presumably).

### CODE REVIEW

**Comment by bshanahan** [comment]
12 as we we'd wsw2

File: include/output.hxx [hunk 1]
```diff
@@ -140,21 +147,22 @@ public:
...
-      : base(base), enabled(base->enabled), base_is_cond(true) {};
+      : base(base), enabled(base->enabled) {};
 
   /// If enabled, writes a string using C printf formatting
   /// by calling base->vwrite
   /// This string is then sent to log file and stdout (on processor 0)
   void write(const char *str, ...) override;
   void vwrite(const char *str, va_list va) override {
     if (enabled) {
+      ASSERT0(base != nullptr);
```
Comment (dschwoerer) [inline include/output.hxx]: Should be ASSERT1 or ASSERT2, as this is never set to a nullptr.
In the Fields we avoid the checks, and in the case of the fields there is actually a fair chance that they have not allocated the data ...
I guess have more well defined definitions of what the levels mean would be nice.
I think level0 should be only some very basic checks, that are fast and also have rather nasty consequences - and nullptr dereference results in a _clean_ segfault ...

File: src/field/field2d.cxx [hunk 2]
```diff
@@ -535,26 +535,36 @@ Field2D pow(BoutReal lhs, const Field2D &rhs, REGION rgn) {
...
-/// Check if the data is valid
-void checkData(const Field2D &f, REGION region) {
-  if(!f.isAllocated()) {
-    throw BoutException("Field2D: Operation on empty data\n");
-  }
-
+// Internal routine to avoid ugliness with interactions between CHECK
+// levels and UNUSED parameters
 #if CHECK > 2
+void checkData_is_finite_on_region(const Field2D &f, REGION region) {
```
Comment (bendudson) [inline src/field/field2d.cxx]: I'm not sure about this; possibly replaces one kind of ugliness with another. Pedantic point: probably should be `checkDataIsFiniteOnRegion`.

Comment (ZedThree) [reply src/field/field2d.cxx]: Good catch on the name, thanks!

Yes, I'm not very happy with it. I think the other alternatives were:

a)
```cpp
#if CHECK > 2
void checkData(const Field2d &f, REGION region)
#elif CHECK > 0
void checkData(const Field2d &f, REGION UNUSED(region))
#endif
{
```
b)
```cpp
void checkData(const Field2d &f, 
#if CHECK > 2
               REGION region)
#elif CHECK > 0
               REGION UNUSED(region))
#endif
{
```
Or of course, just don't warn on unused parameters (by default). It's not really helpful for end-users, but developers should really have it on.

Comment (dschwoerer) [reply src/field/field2d.cxx]: Maybe only not warn about unused parameters when compiled with checks?

**Pull request merged**